bench/workgen/workgen.cxx

```diff
@@ -248,6 +281,231 @@ WorkloadRunner::start_table_idle_cycle(WT_CONNECTION *conn)
     }
     return 0;
 }
+
+/*
+ * Get the cumulative size of all the files under the given directory.
+ */
+static uint32_t
+get_dir_size_mb(std::string dir)
+{
+    uint64_t result = 0;
+    for (const auto &path : std::filesystem::recursive_directory_iterator(dir)) {
+        try {
+            result += std::filesystem::file_size(path);
+        } catch (std::filesystem::filesystem_error &e) {
+            /*
+             * A file might be dropped between listing the directory contents and getting the file
+             * sizes. Ignore such errors.
+             */
+            ASSERT(std::string(e.what()).find("No such file or directory") != std::string::npos);
+            continue;
+        }
+    }
+    return result / (WT_THOUSAND * WT_THOUSAND);
+}
+
+/*
+ * This function will create "dynamic_create_count" tables every "dynamic_create_period" seconds. It
+ * will also monitor the database size and stop creating tables once the size crosses
+ * "dynamic_create_target". It will restart table creation if the database size drops below
+ * "dynamic_create_trigger".
+ */
+int
+WorkloadRunner::start_dynamic_create(WT_CONNECTION *conn)
+{
+    WT_SESSION *session;
+    if (conn->open_session(conn, nullptr, nullptr, &session) != 0) {
+        THROW("Error opening a session.");
+    }
+
+    WT_DECL_RET;
+    ContextInternal *icontext = _workload->_context->_internal;
+    char uri[BUF_SIZE];
+    uint64_t total_creates = 0;
+    uint32_t db_size = get_dir_size_mb(_wt_home);
+
+    // Initially we start creating tables if the database size is less than the create target.
+    bool creating = db_size < _workload->options.dynamic_create_target;
+
+    while (!stopping) {
+        /*
+         * If we have been creating tables, continue until we reach the create target size. If we
+         * have not been creating tables, begin to do so if the database size falls below the create
+         * trigger.
+         */
+        db_size = get_dir_size_mb(_wt_home);
+        if (creating) {
+            creating = db_size < _workload->options.dynamic_create_target;
+            if (!creating) {
+                VERBOSE(*_workload,
+                  "Stopped creating new tables. db_size now "
+                    << db_size << " MB has reached the create target of "
+                    << _workload->options.dynamic_create_target << " MB.");
+            }
+        } else {
+            creating = db_size < _workload->options.dynamic_create_trigger;
+            if (creating) {
+                VERBOSE(*_workload,
+                  "Started creating new tables. db_size now "
+                    << db_size << " MB has reached the create trigger of "
+                    << _workload->options.dynamic_create_trigger << " MB.");
+            }
+        }
+        if (!creating) {
+            sleep(_workload->options.dynamic_create_period);
+            continue;
+        }
+
+        for (int creates = 0; creates < _workload->options.dynamic_create_count;
```

Comment:
We should check for `stopping` here too.

Reply:
We dont really need to stop between a batch of creates. I think this more of stopping between the sleep periods. But I am thinking of an alternate way to structure the code.

Reply:
I am not sure if I fully agree with this, if one wants a test to run for x seconds, it should not depend on what the workload is doing before stopping, I think we should stop as quickly as possible to meet the desired running time. We could document what we decide to do.

For the alternate way, you could simply check if you are supposed to create a table inside the `while` loop to remove all the `for` loops. When you are creating, you cannot delete. If you have created, you have to sleep in between. Once you are no longer created, you check whether you need to drop tables etc.

Reply:
I consider the creation of x tables equivalent to an operation. Let's say this is a modify operation with a throttle, it goes something like:

```
while (!stopping) {
    modify-operation;
    sleep(throttle_ms);
}
```

Now, we don't necessarily abort the modify operation when stopping gets to true, we let that operation complete, we just stopping doing more such operations. Similarly the user has asked to create x tables every y seconds. I prefer to create those x tables as a unit before stopping. If you were to only run creates and not drops for instance, at any point you stop, I expect the database to hence have a multiple of x tables.

Said that, I will restructure the code as part of WT-10467, we can discuss this further as part of that ticket.